typing/typemod.ml

```diff
@@ -3281,14 +3308,33 @@ let report_error ~loc _env = function
         "@[Multiple definition of the %s name %s.@ \
          Names must be unique in a given structure or signature.@]"
         (Sig_component_kind.to_string kind) name
-  | Non_generalizable typ ->
+  | Non_generalizable { vars; expression } ->
+      prepare_for_printing vars;
+      add_type_to_preparation expression;
       Location.errorf ~loc
         "@[The type of this expression,@ %a,@ \
-           contains type variables that cannot be generalized@]" type_scheme typ
-  | Non_generalizable_module mty ->
-      Location.errorf ~loc
+         contains the non-generalizable type variable(s) %a@ %s@]"
+        prepared_type_scheme expression
+        (pp_print_list ~pp_sep:(fun f () -> fprintf f ",@ ")
+           prepared_type_scheme) vars
+        (ref_manual_value_restriction ())
+  | Non_generalizable_module { vars; mty; item } ->
+      prepare_for_printing vars;
+      add_type_to_preparation item.val_type;
+      let sub =
+        [ Location.msg ~loc:item.val_loc
+            "This item has the non-generalizable type@ %a."
```

Comment:
Same question as before: `the non-generalizable type` or `a non-generalizable type`?